testing/trino-product-tests/src/main/java/io/trino/tests/product/jdbc/TestExternalAuthorizerOAuth2RefreshToken.java

```diff
@@ -0,0 +1,166 @@
+    {
+        OkHttpClient.Builder httpClientBuilder = new OkHttpClient.Builder();
+        KeyStore keyStore = KeyStore.getInstance(new File(truststorePath), truststorePassword.toCharArray());
+        HandshakeCertificates.Builder certificatesBuilder = new HandshakeCertificates.Builder();
+        keyStore.aliases().asIterator().forEachRemaining(alias -> {
+            try {
+                Certificate certificate = keyStore.getCertificate(alias);
+                if (certificate instanceof X509Certificate) {
+                    certificatesBuilder.addTrustedCertificate((X509Certificate) certificate);
+                }
+            }
+            catch (KeyStoreException e) {
+                throw new RuntimeException(e);
+            }
+        });
+        HandshakeCertificates certificates = certificatesBuilder.build();
+        httpClientBuilder.sslSocketFactory(certificates.sslSocketFactory(), certificates.trustManager());
+        httpClientBuilder.followRedirects(true);
+        httpClientBuilder.cookieJar(new JavaNetCookieJar(new CookieManager()));
+        httpClient = httpClientBuilder.build();
+    }
+
+    @Test(groups = {OAUTH2_REFRESH, PROFILE_SPECIFIC_TESTS})
+    public void shouldRefreshTokenAfterTokenExpire()
+            throws Exception
+    {
+        prepareHandler();
+        try (Connection connection = DriverManager.getConnection(jdbcUrl);
+                PreparedStatement statement = connection.prepareStatement("SELECT * FROM tpch.tiny.nation");
+                ResultSet results = statement.executeQuery()) {
+            assertThat(forResultSet(results)).matches(TpchTableResults.PRESTO_NATION_RESULT);
+
+            //Wait until the token expires. See: HydraIdentityProvider.TTL_ACCESS_TOKEN_IN_SECONDS
+            SECONDS.sleep(10);
+
+            disableHandler();
+            assertThatNoException().isThrownBy(() -> {
+                try (PreparedStatement repeatedStatement = connection.prepareStatement("SELECT * FROM tpch.tiny.nation");
+                        ResultSet repeatedResults = repeatedStatement.executeQuery()) {
+                    assertThat(forResultSet(repeatedResults)).matches(TpchTableResults.PRESTO_NATION_RESULT);
+                }
+            });
+        }
+    }
+
+    @Test(groups = {OAUTH2_REFRESH, PROFILE_SPECIFIC_TESTS})
+    public void shouldForceTokenRefreshBeforeAccessTokenExpires()
+            throws Exception
+    {
+        prepareHandler();
+        try (Connection connection = DriverManager.getConnection(jdbcUrl);
+                PreparedStatement statement = connection.prepareStatement("SELECT * FROM tpch.tiny.nation");
+                ResultSet results = statement.executeQuery()) {
+            assertThat(forResultSet(results)).matches(TpchTableResults.PRESTO_NATION_RESULT);
+
+            //Wait until the internally issued token expires. See: http-server.authentication.oauth2.refresh-tokens.access-token.expiration
+            SECONDS.sleep(4);
```

Comment:
Is this really testing what it should? I'd say the token didn't expire yet after 4 seconds. The test above seems to confirm that.